### PR TITLE
[DSIP-65] SubWorkflow logic task support pause/kill/recover

### LINKED ISSUE #16480: [DSIP-65] SubWorkflow logic task support failover/repeat running/pause/kill/recover
### Search before asking

- [X] I had searched in the [DSIP](https://github.com/apache/dolphinscheduler/issues/14102) and found no similar DSIP.

### Motivation

Right now the sub workflow doesn't have a good implementation for failover/repeat running/pause/kill/recover, we should redesign this.

### Design Detail

The sub workflow task represent a workflow.
image

When the task running, it will create a sub workflow instance, and track its state.

image

And once we trigger/pause/stop/recover the parent workflow we should trigger/pause/stop/recover the sub workflow instance.

Once the sub workflow task running, it will trigger a sub workflow instance. We will use SubWorkflowLogicTaskRuntimeContext to store the sub workflow instance id, so once the task running, we will get the SubWorkflowLogicTaskRuntimeContext, and using the SubWorkflowLogicTaskRuntimeContext to calculate status.

```java
public class SubWorkflowLogicTaskRuntimeContext {

    private Integer subWorkflowInstanceId;

}
```

When the sub workflow task running, we will trigger the sub workflow by the operation type. 

```java
private SubWorkflowLogicTaskRuntimeContext initializeSubWorkflowInstance() {
        // todo: doFailover if the runtime context is not null and task is generated by failover

        if (subWorkflowLogicTaskRuntimeContext == null) {
            return triggerNewSubWorkflow();
        }

        switch (workflowExecutionRunnable.getWorkflowInstance().getCommandType()) {
            case RECOVER_SUSPENDED_PROCESS:
                return recoverFromSuspendTasks();
            case START_FAILURE_TASK_PROCESS:
                return recoverFromFailedTasks();
            default:
                return triggerNewSubWorkflow();
        }

    }
```

And when we pause/kill the task, we will pause/stop the sub workflow, once the sub workflow instance has been paused/stopped, we can get the status after track status.

```java
@Override
    public void pause() throws MasterTaskExecuteException {
        if (subWorkflowLogicTaskRuntimeContext == null) {
            log.info("subWorkflowLogicTaskRuntimeContext is null cannot pause");
            return;
        }
        final Integer subWorkflowInstanceId = subWorkflowLogicTaskRuntimeContext.getSubWorkflowInstanceId();
        final WorkflowInstancePauseResponse pauseResponse = applicationContext
                .getBean(SubWorkflowControlClient.class)
                .pauseWorkflowInstance(new WorkflowInstancePauseRequest(subWorkflowInstanceId));
        if (pauseResponse.isSuccess()) {
            log.info("Pause sub workflowInstance: id={}", subWorkflowInstanceId + " success");
        } else {
            log.info("Pause sub workflowInstance: id={} failed with response: {}", subWorkflowInstanceId,
                    pauseResponse);
        }
    }

    @Override
    public void kill() throws MasterTaskExecuteException {
        if (subWorkflowLogicTaskRuntimeContext == null) {
            log.info("subWorkflowLogicTaskRuntimeContext is null cannot kill");
            return;
        }
        final Integer subWorkflowInstanceId = subWorkflowLogicTaskRuntimeContext.getSubWorkflowInstanceId();
        final WorkflowInstanceStopResponse stopResponse = applicationContext
                .getBean(SubWorkflowControlClient.class)
                .stopWorkflowInstance(new WorkflowInstanceStopRequest(subWorkflowInstanceId));
        if (stopResponse.isSuccess()) {
            log.info("Kill sub workflowInstance: id={}", subWorkflowInstanceId + " success");
        } else {
            log.info("Kill sub workflowInstance: id={} failed with response: {}", subWorkflowInstanceId, stopResponse);
        }
    }
```

### Compatibility, Deprecation, and Migration Plan

Compatibility with old version.

### Test Plan

Test by Master IT.

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by ruanwenjun**
## Purpose of the pull request

close https://github.com/apache/dolphinscheduler/issues/16480

## Brief change log

- Sub workflow task add implementation
- Add sub workflow IT
- Change WorkflowGraphBfsVisitor to WorkflowGraphTopologyLogicalVisitor

## Verify this pull request

This pull request is code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

(or)

## Pull Request Notice
[Pull Request Notice](https://github.com/apache/dolphinscheduler/blob/dev/docs/docs/en/contribute/join/pull-request.md)

If your pull request contain incompatible change, you should also add it to `docs/docs/en/guide/upgrede/incompatible.md`

### CODE REVIEW

**Review by SbloodyS — APPROVED** [review]
LGTM.

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/task/dependent/DependentLogicTask.java [hunk 1]
```diff
@@ -72,34 +70,24 @@ public DependentLogicTask(TaskExecutionContext taskExecutionContext,
...
-                .getTaskInstance();
-        if (taskInstance == null) {
-            log.error("Cannot find the TaskInstance in workflowExecuteRunnable");
-            return;
-        }
-        taskInstance.setState(TaskExecutionStatus.PAUSE);
-        taskInstance.setEndTime(new Date());
-        taskInstanceDao.upsertTaskInstance(taskInstance);
-        super.pause();
+        // todo: support pause
```
Comment (caishunfeng) [inline dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/task/dependent/DependentLogicTask.java:85]: Dependent Task is very important in ds, so I think this is a high priority todo.

If it cannot be completed in a short time, I suggest to add some warn logs here.

**Review by caishunfeng — APPROVED** [review]
LGTM

File: dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/task/dependent/DependentLogicTask.java (hunk 1 above)
Comment (ruanwenjun) [reply dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/runner/task/dependent/DependentLogicTask.java:85]: Yes, will use https://github.com/apache/dolphinscheduler/issues/16577 to fix this.

**Pull request merged**